### PR TITLE
Add basic user management

### PR DESCRIPTION
**Pull request by Zlopez**
Changes:
* Adds new page `/users` which is similar to `/flags` page.
  This user management will help promote new admins or ban users.
screenshot from 2018-11-16 11-22-40
* Add `admin` flag to `User` model.
* User property `admin` is renamed to `is_admin`.
* Admin flag for preconfigured users is set on `is_admin` call.

Signed-off-by: Michal Konečný <mkonecny@redhat.com>

### CODE REVIEW

File: anitya/admin.py [hunk 1]
```diff
@@ -464,3 +464,194 @@ def set_flag_state(flag_id, state):
     return flask.redirect(
         flask.url_for('anitya_ui.browse_flags')
     )
+
+
+@ui_blueprint.route('/users', methods=['GET'])
+@login_required
+def browse_users():
+
+    if not is_admin():
```
Comment (cverna) [inline anitya/admin.py]: It would be nice to refactor this code in a decorator, similar to how `@login_required` works. But this should be done in another PR.

File: anitya/admin.py [hunk 2]
```diff
@@ -464,3 +464,194 @@ def set_flag_state(flag_id, state):
...
+
+
+@ui_blueprint.route('/users/<user_id>/admin/<state>', methods=['POST'])
+@login_required
+def set_user_admin_state(user_id, state):
+
+    if not is_admin():
+        flask.abort(401)
+
+    if state.upper() == "TRUE":
```
Comment (cverna) [inline anitya/admin.py]: It might be better to use a form here with some boolean fields , what do you think ?

File: anitya/admin.py [hunk 3]
```diff
@@ -464,3 +464,194 @@ def set_flag_state(flag_id, state):
...
+
+
+@ui_blueprint.route('/users/<user_id>/active/<state>', methods=['POST'])
+@login_required
+def set_user_active_state(user_id, state):
+
+    if not is_admin():
+        flask.abort(401)
+
+    if state.upper() == "TRUE":
```
Comment (cverna) [inline anitya/admin.py]: Same as above maybe using a form would make sense here

File: anitya/db/migrations/versions/b13662e5d288_add_admin_flag.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+def upgrade():
+    """ Add 'admin' flag to users table. """
+    op.add_column(
+        'users',
+        sa.Column(
+            'admin',
+            sa.Boolean,
+            default=False)
+    )
```
Comment (cverna) [inline anitya/db/migrations/versions/b13662e5d288_add_admin_flag.py]: This will remove the admin permission to the users that are currently admin. You might want to make sure that the current admins have the value set to `True` during the migration.

File: anitya/db/models.py [hunk 5]
```diff
@@ -896,16 +898,21 @@ class User(Base):
...
-        Determine if this user is an administrator.
+        Determine if this user is an administrator. Set admin flag
+        if the user is preconfigured.
 
         Returns:
             bool: True if the user is an administrator.
         """
-        return six.text_type(self.id) in anitya_config.get('ANITYA_WEB_ADMINS', [])
+        if not self.admin:
+            if six.text_type(self.id) in anitya_config.get('ANITYA_WEB_ADMINS', []):
```
Comment (cverna) [inline anitya/db/models.py]: I think that it would be good to drop support for this config key. That way there would be only one place where you can manage admins.

File: anitya/templates/users.html [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        <td>{{ user.email }}</td>
+        <td>{{ user.is_admin }}</td>
+        <td>{{ user.is_active }}</td>
+        <td>
+            {% if user.is_admin %}
+            <form method="POST" action="{{ url_for('anitya_ui.set_user_admin_state',
+                user_id=user.id, state=False) }}">
+                <button type="submit" class="btn btn-danger btn-sm pull-right inline"
+                        id="user_btn">
+                    Revoke admin
```
Comment (cverna) [inline anitya/templates/users.html]:
```suggestion
                    Remove admin permissions
```

File: anitya/templates/users.html [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+                </button>
+                {% else %}
+                <form method="POST" action="{{ url_for('anitya_ui.set_user_admin_state',
+                user_id=user.id, state=True) }}">
+                    <button type="submit" class="btn btn-success btn-sm pull-right inline"
+                            id="user_btn">
+                        Promote admin
```
Comment (cverna) [inline anitya/templates/users.html]:
```suggestion
                    Give admin permissions
```

File: anitya/templates/users.html [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+                <button type="submit" class="btn btn-danger btn-sm pull-right inline"
+                        id="user_btn">
+                    Ban
+                </button>
+                {% else %}
+                <form method="POST" action="{{ url_for('anitya_ui.set_user_active_state',
+                user_id=user.id, state=True) }}">
+                    <button type="submit" class="btn btn-success btn-sm pull-right inline"
+                            id="user_btn">
+                        Remove ban
```
Comment (cverna) [inline anitya/templates/users.html]:
```suggestion
                    Remove ban
```

File: anitya/tests/test_flask_admin.py [hunk 9]
```diff
@@ -803,3 +805,563 @@ def test_set_flag(self):
...
+
+    def test_missing_state(self):
+        """Assert an missing state results in HTTP 404."""
+        with login_user(self.flask_app, self.admin):
+            output = self.client.post('/users/{}/admin/'.format(
+                six.text_type(self.user.id)
+            ))
+            self.assertEqual(404, output.status_code)
+
+    def test_missing(self):
```
Comment (cverna) [inline anitya/tests/test_flask_admin.py]:
```suggestion
    def test_missing_user(self):
```

File: anitya/tests/test_flask_admin.py [hunk 10]
```diff
@@ -803,3 +805,563 @@ def test_set_flag(self):
...
+
+    def test_missing_state(self):
+        """Assert an missing state results in HTTP 404."""
+        with login_user(self.flask_app, self.admin):
+            output = self.client.post('/users/{}/active/'.format(
+                six.text_type(self.user.id)
+            ))
+            self.assertEqual(404, output.status_code)
+
+    def test_missing(self):
```
Comment (cverna) [inline anitya/tests/test_flask_admin.py]:
```suggestion
    def test_missing_user(self):
```

File: anitya/admin.py (hunk 1 above)
Comment (Zlopez) [reply anitya/admin.py]: This could be nice. I will add it to my TODO list.

File: anitya/admin.py (hunk 2 above)
Comment (Zlopez) [reply anitya/admin.py]: There is a form already on the page, but this change is done by button which sets the state value.

File: anitya/db/migrations/versions/b13662e5d288_add_admin_flag.py (hunk 4 above)
Comment (Zlopez) [reply anitya/db/migrations/versions/b13662e5d288_add_admin_flag.py]: Current admins are defined in the configuration file, so this shouldn't change anything.

File: anitya/db/models.py (hunk 5 above)
Comment (Zlopez) [reply anitya/db/models.py]: The admins set in the configuration file shouldn't be stripped of this privilege. This is failsafe against some admin going rampage and revoking rights for everyone else and yourself. Also the configuration file could be changed in ansible, so you can easily remove or add admin even if you don't have the admin rights in the applications itself.

File: anitya/templates/users.html (hunk 6 above)
Comment (Zlopez) [reply anitya/templates/users.html]: I need to check, how this will actually look on the page first.

File: anitya/admin.py (hunk 2 above)
Comment (cverna) [reply anitya/admin.py]: Ack

File: anitya/db/models.py (hunk 5 above)
Comment (cverna) [reply anitya/db/models.py]: Ok, I was just concerned that with time it will  not be obvious where the admins are managed (ie config or in the db).

File: anitya/admin.py (hunk 3 above)
Comment (Zlopez) [reply anitya/admin.py]: Same as above

**Comment by Zlopez** [comment]
@cverna 
Addressed your suggestions and rebased.
